sealedenumprocessor/src/main/kotlin/com/livefront/sealedenum/internal/SealedEnumTypeSpec.kt

```diff
@@ -91,4 +109,114 @@ internal data class SealedEnumTypeSpec(
+                }
+                addStatement(
+                    "else -> throw IllegalArgumentException(%P)",
+                    "No sealed enum constant \$name"
+                )
+                endControlFlow()
+            }
+        }
+        .build()
+
+    private fun createSealedObjectToEnumFunction(enumForSealedEnum: ClassName): FunSpec =
+        FunSpec.builder("sealedObjectToEnum")
+            .addParameter("obj", parameterizedSealedClass)
+            .addModifiers(KModifier.OVERRIDE)
+            .returns(enumForSealedEnum)
+            .apply {
+                if (sealedObjects.isEmpty()) {
+                    addStatement(
+                        "throw AssertionError(%S)",
+                        "Constructing a ${sealedClass.simpleName} is impossible, since it has no sealed subclasses"
+                    )
+                } else {
+                    beginControlFlow("return when (obj)")
+                    sealedObjects.forEach { obj ->
+                        addStatement(
+                            "%T -> %T",
+                            obj,
+                            enumForSealedEnum.nestedClass(obj.simpleNames.joinToString("_"))
+                        )
+                    }
+                    endControlFlow()
+                }
+            }
+            .build()
+
+    private fun createEnumToSealedObjectFunction(enumForSealedEnum: ClassName): FunSpec =
+        FunSpec.builder("enumToSealedObject")
+            .addParameter("enum", enumForSealedEnum)
+            .addModifiers(KModifier.OVERRIDE)
+            .returns(parameterizedSealedClass)
+            .apply {
+                if (sealedObjects.isEmpty()) {
+                    addStatement(
+                        "throw AssertionError(%S)",
+                        "Constructing a ${sealedClass.simpleName} is impossible, since it has no sealed subclasses"
+                    )
+                } else {
+                    beginControlFlow("return when (enum)")
+                    sealedObjects.forEach { obj ->
+                        addStatement(
+                            "%T -> %T",
+                            enumForSealedEnum.nestedClass(obj.createValueName()),
+                            obj
+                        )
+                    }
+                    endControlFlow()
+                }
+            }
+            .build()
+
+    private fun createEnumClassGetter(enumForSealedEnum: ClassName): PropertySpec {
+        val parameterizedClass = Class::class.asClassName().parameterizedBy(enumForSealedEnum)
+
+        return PropertySpec.builder("enumClass", parameterizedClass)
+            .addModifiers(KModifier.OVERRIDE)
+            .getter(
+                FunSpec.getterBuilder()
+                    .addStatement("return %T::class.java", enumForSealedEnum)
+                    .build()
+            )
+            .build()
+    }
```

Comment:
My brain kind of shuts off when I see this much KotlinPoet so I'll take your word for it...